bench/Autofac.Benchmarks/Decorators/DecoratorBenchmarkBase.cs

```diff
@@ -9,17 +9,33 @@ public abstract class DecoratorBenchmarkBase<TCommandHandler>
         protected IContainer Container { get; set; }
 
         [Benchmark]
-        public virtual void EnumerableResolve()
+        [Arguments(1)]
+        [Arguments(2)]
+        [Arguments(3)]
```

Comment:
I expanded on all of the decorator benchmarks to have repetition runs, since the first execution of Resolve<> is always really fast, but seeing the effect of repeated resolutions when using shared or independent resolutions provided some good data. My BenchmarkDotNet experience is literally only these files, so if there are recommendations for improvement, I'm all ears!
Sample output with repetitions for KeylessNestedSharedInstanceLambdaBenchmark:
![image](https://user-images.githubusercontent.com/18561847/81252599-2899cf80-8fdb-11ea-8571-3e4d598ba64a.png)

Reply:
Just so you are aware, BenchmarkDotNet outputs a markdown file that contains GitHub-compatible markdown you can just paste straight here (rather than needing an image).

Reply:
The problem with the repetition approach you have here is that the benchmark methods are already being executed millions of times as it is, against a single ``Container`` instance, to get the average time. See the comment on the repetition logic as to why you are actually seeing a discrepancy between 1 and 2.

So, 99.999999% of runs are going to be using the shared instance.

This sort of global state behaviour with benchmarks is always fun. In this particular case, my suggestion to you would be to resolve from a lifetime scope, and use a baseline to determine the cost of creating the lifetime scope.

Something like:

1. Change the shared scope of the benchmarks to be lifetime scope, rather than SingleInstance.
2. Derive from a new base class ``ScopedDecoratorBenchmarkBase`` (or something).
3. In this benchmark class, create a new lifetime scope at the beginning of each benchmark, and resolve from that.
4. Add a new [baseline](https://benchmarkdotnet.org/articles/features/baselines.html) benchmark in your new base class that's only job is to create and dispose of a lifetime scope. This gives us the 'no-op' cost of the **actual** benchmarks.

Hope that was clear, I really like benchmarks, but they can be a bit black-magic sometimes.

Reply:
The suggestion is completely clear! I misunderstood how global "GlobalSetup" really was, so I see how that + repetition didn't quite do what I was thinking it was doing. I was coming at it from the perspective of the fact that a single call to Resolve<>() doesn't really expose the costs of just the instance lookup when the initial resolve created a shared instance. With your suggestion about creating a lifetime scope, I'm assuming the repetition element would still be valuable, since it would expose the costs of instance lookup vs instance creation?

Reply:
Correct, the repetition would still be valuable in that case, because you'd be resolving X times from the lifetime scope before it is disposed.

I'd _expect_ to see the highest cost on the first resolve, then lower on the second, and all subsequent repetitions should be of similar cost to the second.